File: common/stats.rs
use serde::Serialize;

#[derive(Serialize)]
pub struct StatsData {
    pub mean_ns: f64,
    pub stddev_ns: f64,
    pub min_ns: f64,
    pub max_ns: f64,
    pub p50_ns: f64,
    pub p95_ns: f64,
    pub p99_ns: f64,
    pub throughput_ops_sec: f64,
    pub throughput_mb_sec: f64,
}

pub fn calculate_stats(mut times_ns: Vec<u64>, total_bytes: usize) -> StatsData {
    if times_ns.is_empty() {
        return StatsData {
            mean_ns: 0.0, stddev_ns: 0.0, min_ns: 0.0, max_ns: 0.0,
            p50_ns: 0.0, p95_ns: 0.0, p99_ns: 0.0, throughput_ops_sec: 0.0, throughput_mb_sec: 0.0
        };
    }

    times_ns.sort_unstable();
    let n = times_ns.len() as f64;

    let sum_ns: u64 = times_ns.iter().sum();
    let mean_ns = sum_ns as f64 / n;

    let variance_ns = times_ns.iter()
        .map(|&x| (x as f64 - mean_ns).powi(2))
        .sum::<f64>() / n;
    let stddev_ns = variance_ns.sqrt();

    let total_time_s = sum_ns as f64 / 1e9;

    StatsData {
        mean_ns,
        stddev_ns,
        min_ns: times_ns[0] as f64,
        max_ns: *times_ns.last().unwrap() as f64,
        p50_ns: times_ns[(n * 0.50) as usize] as f64,
        p95_ns: times_ns[(n * 0.95) as usize] as f64,
        p99_ns: times_ns[(n * 0.99) as usize] as f64,
        throughput_ops_sec: if total_time_s > 0.0 { n / total_time_s } else { 0.0 },
        throughput_mb_sec: if total_time_s > 0.0 { (total_bytes as f64 / (1024.0 * 1024.0)) / total_time_s } else { 0.0 }
    }
}

pub fn print_stats(queue_name: &str, stats: &StatsData) {
    let json = serde_json::json!({
        queue_name: stats
    });
    println!("{}", serde_json::to_string_pretty(&json).unwrap());
}

File: common/stats.c
#include "stats.h"
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

int compare_ull(const void *a, const void *b) {
    unsigned long long ua = *(const unsigned long long *)a;
    unsigned long long ub = *(const unsigned long long *)b;
    if (ua < ub) return -1;
    if (ua > ub) return 1;
    return 0;
}

StatsData calculate_stats(unsigned long long *times_ns, size_t n, size_t total_bytes) {
    StatsData stats = {0};
    if (n == 0) return stats;

    qsort(times_ns, n, sizeof(unsigned long long), compare_ull);

    unsigned long long sum_ns = 0;
    for (size_t i = 0; i < n; i++) sum_ns += times_ns[i];

    stats.mean_ns = (double)sum_ns / n;

    double variance_ns = 0;
    for (size_t i = 0; i < n; i++) {
        double diff = times_ns[i] - stats.mean_ns;
        variance_ns += diff * diff;
    }
    variance_ns /= n;
    stats.stddev_ns = sqrt(variance_ns);

    stats.min_ns = times_ns[0];
    stats.max_ns = times_ns[n - 1];
    stats.p50_ns = times_ns[(size_t)(n * 0.50)];
    stats.p95_ns = times_ns[(size_t)(n * 0.95)];
    stats.p99_ns = times_ns[(size_t)(n * 0.99)];

    double total_time_s = sum_ns / 1e9;
    if (total_time_s > 0) {
        stats.throughput_ops_sec = n / total_time_s;
        stats.throughput_mb_sec = ((double)total_bytes / (1024.0 * 1024.0)) / total_time_s;
    }

    return stats;
}

void print_stats(const char *queue_name, StatsData stats) {
    printf("{\n");
    printf("  \"%s\": {\n", queue_name);
    printf("    \"mean_ns\": %f,\n", stats.mean_ns);
    printf("    \"stddev_ns\": %f,\n", stats.stddev_ns);
    printf("    \"min_ns\": %f,\n", stats.min_ns);
    printf("    \"max_ns\": %f,\n", stats.max_ns);
    printf("    \"p50_ns\": %f,\n", stats.p50_ns);
    printf("    \"p95_ns\": %f,\n", stats.p95_ns);
    printf("    \"p99_ns\": %f,\n", stats.p99_ns);
    printf("    \"throughput_ops_sec\": %f,\n", stats.throughput_ops_sec);
    printf("    \"throughput_mb_sec\": %f\n", stats.throughput_mb_sec);
    printf("  }\n");
    printf("}\n");
}
